Support local primary-replica-only for native restore

robertbindar opened the pull request:
This PR extends the restore API so that it accepts primary_replica_only as parameter and it combines the concepts of primary-replica-only with scoped streaming so that with:
- `scope=all primary_replica_only=true` The restoring node will stream to the global primary replica only
- `scope=dc primary_replica_only=true` The restoring node will stream to the local primary replica only.
- `scope=rack primary_replica_only=true` The restoring node will stream only to the primary replica from within its own rack (with rf=#racks, the restoring node will stream only to itself)
- `scope=node primary_replica_only=true` is not allowed, the restoring node will always stream only to itself so the primary_replica_only parameter wouldn't make sense.

The PR also adjusts the `nodetool refresh` restriction on running restore with both primary_replica_only and scope, it adds primary_replica_only to `nodetool restore` and it adds cluster tests for primary replica within scope.


Fixes #26584

bhalevy commented on api/api-doc/storage_service.json:
```diff
@@ -942,6 +942,14 @@
+                      },
+                      {
+                         "name":"primary_replica_only",
+                         "description":"Load the sstables and stream to only local primary replica node.",
```
Also the patch title: "Support primary_replica_only for native restore API" is overreaching.
I'd tune it down to something like "api: storage_service/restore: add primary_replica_only option"

regevran replied:
Primary replica and scope are mutually exclusive. 
One cannot use both in the same request.

Michal-Leszczynski replied:
I agree that it would be cleaner to either have 2 additional scope options, or make it work with scope (and describe this interaction). 
In the current state, it's not really clear to me whether `primary_replica_only` works on cluster or dc primary replica, but I guess that's consistent with the `/storage_service/sstables/{keyspace}` api where I'm also not sure which one is used (I guess the cluster one).

robertbindar commented:
The latest version:
- combines the concepts of primary-replica-only with scope:
```
- scope=all primary_replica_only=true gets the global primary replica
- scope=dc primary_replica_only=true gets the local primary replica
- scope=rack primary_replica_only=true is like a noop, it gets the only
  replica in the rack (rf=#racks)
- scope=node primary_replica_only=node is not allowed
```
- adjusts the exception and tests for `nodetool refresh`
- adds the primary-replica-option in `nodetool restore`
- TODO: work in progress to add a python test

robertbindar commented on sstables_loader.cc:
```diff
@@ -223,6 +220,11 @@ host_id_vector_replica_set sstable_streamer::get_endpoints(const dht::token& tok
     }) | std::ranges::to<host_id_vector_replica_set>();
+
+    if (_primary_replica_only && !filtered_endpoints.empty()) {
+        filtered_endpoints.resize(1);
```
> Please review #26385 and see if we have an opportunity to share some of the code

yes, I listed it as a dependency for this PR because it does exactly what we need here for the tablets case, it pushes down the filter by scope so that in the end the primary replica will be: `scope_filter(replicas)[tablet_id % num_filtered_replicas]`

robertbindar replied:
> > > Also, what about replicas in wrote_both state?
> > 
> > 
> > The way write_both it'll work after PR #26385 is that both previous and next replica _within the scope_ will be selected. It looks from the code that in this case, the tablet `get_primary_replicas` function may return 2 primary replicas if they both are within the passed scope (most likely scope=dc is the only possible case here).
> 
> Thinking again about this, when primary-replica-only=true we require repair to run after streaming anyway, so there is really no need to stream to the pending replica. If primary-replica-only=false and there is a pending replica then it should be included depending on the scope: all/dc/rack vs. the location of the pending replica. As for scope=node, it's a good question. It's always safe to stream to the pending replica, but it will create space amplification, e.g. for tablet-aware restore if we allow migration to happen in parallel. If we can determine that this node is the source for migration, i.e. it is in the "pre" replica set but not in the "post" replica set, we should definitely stream to both this node and the pending replica.

But I thought it's settled that scope=node is not allowed with primary_replica_only=true.

bhalevy replied:
True.  I referred to in the case of primary_replica_only=false

regevran commented on sstables_loader.cc:
```diff
@@ -154,9 +154,6 @@ class sstable_streamer {
             , _unlink_sstables(unlink)
             , _stream_scope(scope)
     {
-        if (_primary_replica_only && _stream_scope != stream_scope::all) {
```
> I would expect that it would save the mutations to local nodes provided this node is primary replica

Yes. Nodes which are in the "intersection" of the two groups: {`primary-replica-only`} ^ {`scope`} should be considered.
{`primary-replica-only`} = {true, false}
{`scope`} = {all, dc, rack, node}

robertbindar replied:
At some point I added https://github.com/scylladb/scylladb/pull/26609#issuecomment-3432210450 but it's easy to miss. I'll update the PR cover with the description.
In short, scope=node is not allowed with primary replica, it'd stream only to the restoring node, so primary_replica_only parameter wouldn't make any difference. With scope=rack, the selection of primary replica will be made from the list of replicas within the rack (with rf=#racks it'd also mean it'll stream only to the restoring node, but this option should be allowed as long as rf=#racks is not mandatory).

robertbindar commented on sstables_loader.cc:
```diff
@@ -211,7 +208,7 @@ class tablet_sstable_streamer : public sstable_streamer {
 };
 
 host_id_vector_replica_set sstable_streamer::get_endpoints(const dht::token& token) const {
-    return get_all_endpoints(token) | std::views::filter([&topo = _erm->get_topology(), scope = _stream_scope] (const auto& ep) {
```
it's rebased now, not applicable, the previous version was based on a poor understanding on my side of how that code works

xemul commented:
It's a preexisting problem (but maybe it relates to [this](https://github.com/scylladb/scylladb/pull/26609/commits/4e8b9f28b98666b4bc47f3f9ee89199f9716ff00) newbie) but still I want to raise it here.

The tablet_sstable_streamer::get_primary_endpoints() calls locator::get_primary_replicas(). Does the latter guarantee that _different_ nodes calling it (with coordinated tablet map) will get the same list of nodes? In other words -- the "replicas" field in the system.tablets table is a set, but is this set "sorted" similarly on different nodes, or does it depend on the order of mutations applied to it? (@bhalevy)

xemul commented on sstables_loader.cc:
```diff
@@ -557,15 +557,15 @@ future<locator::effective_replication_map_ptr> sstables_loader::await_topology_q
 
 future<> sstables_loader::load_and_stream(sstring ks_name, sstring cf_name,
-        ::table_id table_id, std::vector<sstables::shared_sstable> sstables, bool primary, bool unlink, stream_scope scope,
+        ::table_id table_id, std::vector<sstables::shared_sstable> sstables, primary_replica_only primary, bool unlink, stream_scope scope,
```
This concern still stands

robertbindar commented:
> > > It's a preexisting problem (but maybe it relates to [this](https://github.com/scylladb/scylladb/pull/26609/commits/4e8b9f28b98666b4bc47f3f9ee89199f9716ff00) newbie) but still I want to raise it here.
> > > The tablet_sstable_streamer::get_primary_endpoints() calls locator::get_primary_replicas(). Does the latter guarantee that _different_ nodes calling it (with coordinated tablet map) will get the same list of nodes? In other words -- the "replicas" field in the system.tablets table is a set, but is this set "sorted" similarly on different nodes, or does it depend on the order of mutations applied to it? (@bhalevy)
> > 
> > 
> > Modulo tablet splits/merges (as you said: "with coordinated tablet map), different nodes will return the same primary replica as the "sets" they are working on are a small_vector ordered the same way everywhere, as populated from system.tablets, in:
> > ```c++
> > static
> > tablet_replica_set deserialize_replica_set(cql3::untyped_result_set_row::view_type raw_value) {
> >     return tablet_replica_set_from_cell(
> >             replica_set_type->deserialize_value(raw_value));
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The reason I added [`4e8b9f2` (#26609)](https://github.com/scylladb/scylladb/pull/26609/commits/4e8b9f28b98666b4bc47f3f9ee89199f9716ff00) is that I was observed the `replicas` field was _not_ sorted.
> 
> Why does it need to be sorted? I believe that the replicas are currently ordered by dc and rack, so that the selection or primary replica by tablet id will go over all racks in round robin. Sorting by host_id would scramble things around. I think that for a large number of tablets it won't matter much, but since it's not required IMO I wouldn't do it.

@bhalevy in my case I noticed that on the restoring node, the replicas would have different order for different calls of `get_endpoints()`, so for a set of tablet_ids, we ended up selecting the same node multiple times.
Look at `test_restore_primary_replica_different_rack_scope_dc` test in this PR. There are 2 nodes, 2 racks, same dc, rf=2. For some `get_endpoints()` calls, replicas would look like: `[815312ae-51ce-444a-8341-cf5fb1103e48:1, cad4884a-2db4-4d4b-b67d-8b64bb81623d:1]` whilst for other it'll look like: `[cad4884a-2db4-4d4b-b67d-8b64bb81623d:0, 815312ae-51ce-444a-8341-cf5fb1103e48:0]`.
Because of this, for the particular dataset of the test, `tablet_id % replicas.size()` would end up *always* picking node `815312ae-51ce-444a-8341-cf5fb1103e48` and never node `cad4884a-2db4-4d4b-b67d-8b64bb81623d`.
Sorting the replicas vector would let `%` do its uniform distribution job.

robertbindar commented on locator/tablets.cc:
yes, makes sense

robertbindar commented:
> In
> 
> https://github.com/scylladb/scylladb/blob/4e8b9f28b98666b4bc47f3f9ee89199f9716ff00/locator/tablets.cc#L268
> 
> 
> shouldn't we add a pending replica only if the selected primary is not in the `next` set?, and in this case, if the number of replicad in next is the same as in the current replica sey, then look for the replica in next that is not in current (not based on calling maybe_get_primary_replica again).?

I don't understand why would do this, I think this is something to solve in a different issue though. Is it directly related and I don't see it?

robertbindar commented:
> > > > > > > It's a preexisting problem (but maybe it relates to [this](https://github.com/scylladb/scylladb/pull/26609/commits/4e8b9f28b98666b4bc47f3f9ee89199f9716ff00) newbie) but still I want to raise it here.
> > > > > > > The tablet_sstable_streamer::get_primary_endpoints() calls locator::get_primary_replicas(). Does the latter guarantee that _different_ nodes calling it (with coordinated tablet map) will get the same list of nodes? In other words -- the "replicas" field in the system.tablets table is a set, but is this set "sorted" similarly on different nodes, or does it depend on the order of mutations applied to it? (@bhalevy)
> > > > > > 
> > > > > > 
> > > > > > Modulo tablet splits/merges (as you said: "with coordinated tablet map), different nodes will return the same primary replica as the "sets" they are working on are a small_vector ordered the same way everywhere, as populated from system.tablets, in:
> > > > > > ```c++
> > > > > > static
> > > > > > tablet_replica_set deserialize_replica_set(cql3::untyped_result_set_row::view_type raw_value) {
> > > > > >     return tablet_replica_set_from_cell(
> > > > > >             replica_set_type->deserialize_value(raw_value));
> > > > > > }
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > The reason I added [`4e8b9f2` (#26609)](https://github.com/scylladb/scylladb/pull/26609/commits/4e8b9f28b98666b4bc47f3f9ee89199f9716ff00) is that I was observed the `replicas` field was _not_ sorted.
> > > > > 
> > > > > 
> > > > > Why does it need to be sorted? I believe that the replicas are currently ordered by dc and rack, so that the selection or primary replica by tablet id will go over all racks in round robin. Sorting by host_id would scramble things around. I think that for a large number of tablets it won't matter much, but since it's not required IMO I wouldn't do it.
> > > > 
> > > > 
> > > > @bhalevy in my case I noticed that on the restoring node, the replicas would have different order for different calls of `get_endpoints()`, so for a set of tablet_ids, we ended up selecting the same node multiple times. Look at `test_restore_primary_replica_different_rack_scope_dc` test in this PR. There are 2 nodes, 2 racks, same dc, rf=2. For some `get_endpoints()` calls, replicas would look like: `[815312ae-51ce-444a-8341-cf5fb1103e48:1, cad4884a-2db4-4d4b-b67d-8b64bb81623d:1]` whilst for other it'll look like: `[cad4884a-2db4-4d4b-b67d-8b64bb81623d:0, 815312ae-51ce-444a-8341-cf5fb1103e48:0]`. Because of this, for the particular dataset of the test, `tablet_id % replicas.size()` would end up _always_ picking node `815312ae-51ce-444a-8341-cf5fb1103e48` and never node `cad4884a-2db4-4d4b-b67d-8b64bb81623d`. Sorting the replicas vector would let `%` do its uniform distribution job.
> > > 
> > > 
> > > Hmm. If this is the case, then sorting is justified. But that said, let's sort by <dc, rack, host_id> (though we're not supposed to have multiple live replicas in a single rack anymore with ks-rf-valid)
> > 
> > 
> > @bhalevy can you please explain why we need sorting by <dc,rack,host>? I don't understand why and how this helps,
> 
> @robertbindar I prefer the round robin rotation of primary replicas across racks in a dc (per tablet) to be stable when a tablet is migrated (we only migrate tablets within the same rack) and not just be random and possibly suffer from rack imbalance. This will probably become more important with strong data consistency when we'd want to use the local primary replicas as the per-tablet raft group leaders.

Addressed in the last version, thanks Benny.

bhalevy commented on locator/tablets.cc:
```diff
@@ -555,14 +556,26 @@ dht::token_range tablet_map::get_token_range_after_split(const token& t) const n
+        if (loc_a.rack != loc_b.rack) {
+            return loc_a.rack < loc_b.rack;
+        }
+        return a.host < b.host;
```
let's add endpoint_dc_rack::operator<=>(const endpoint_dc_rack&) and simply use projection here.

robertbindar replied:
I'd need also the host_id to make the comparison which `endpoint_dc_rack` doesn't seem to contain.

bhalevy replied:
Hmm, in practice, with rf-rack-valid we won't have multiple hosts in a rack, but that's right in principle.

In any case, I prefer ti have the comparison function defined as a method rather than open coding it here.

robertbindar replied:
added in the last version, thanks.

bhalevy commented on test/pylib/rest_client.py:
```diff
@@ -374,13 +374,14 @@ async def backup(self, node_ip: str, ks: str, table: str, tag: str, dest: str, b
                   "bucket": bucket,
-                  "prefix": prefix}
+                  "prefix": prefix,
+                  "primary_replica_only": "true" if primary_replica_only else "false"}
```
nit: "primary_replica_only": str(primary_replica_only).lower()

robertbindar replied:
won't work, `None` is an option too.

denesb commented:
I still see an unresolved comment: https://github.com/scylladb/scylladb/pull/26609#discussion_r2502793893.
